demos/guide/tls-client-block.c

```diff
@@ -0,0 +1,252 @@
+ */
+
+/*
+ * NB: Changes to this file should also be reflected in
+ * doc/man7/guide-tls-client-block.pod
+ */
+
+#include <string.h>
+
+/* Include the appropriate header file for AF_INET and SOCK_STREAM */
+#ifdef _WIN32 /* Windows */
+# include <winsock2.h>
+#else /* Linux/Unix */
+# include <sys/socket.h>
+#endif
+
+#include <openssl/bio.h>
+#include <openssl/ssl.h>
+#include <openssl/err.h>
+
+/* Helper function to create a BIO connected to the server */
+static BIO *create_socket_bio(const char *hostname, const char *port)
+{
+    int sock;
+    BIO_ADDRINFO *ai = NULL;
+    BIO *bio;
+
+    /*
+     * Create a TCP socket. We could equally use non-OpenSSL calls such as
+     * "socket" here for this and the subsequent lookup, connect and close
+     * functions. But for portability reasons and also so that we get errors
+     * on the OpenSSL stack in the event of a failure we use OpenSSL's
+     * versions of these functions.
+     */
+    sock = BIO_socket(AF_INET, SOCK_STREAM, 0, 0);
+    if (sock == -1)
+        return NULL;
+
+    /*
+     * Lookup IP address info for the server. For simplicity we only look at
+     * the first address returned.
+     */
+    if (!BIO_lookup_ex(hostname, port, BIO_LOOKUP_CLIENT, AF_INET,
+                       SOCK_STREAM, 0, &ai)) {
+        BIO_closesocket(sock);
+        return NULL;
+    }
+
+    /* Connect the socket to the server's address */
+    if (!BIO_connect(sock, BIO_ADDRINFO_address(ai), BIO_SOCK_NODELAY)) {
+        BIO_closesocket(sock);
+        BIO_ADDRINFO_free(ai);
+        return NULL;
+    }
+
+    /* Free the address information resources we allocated earlier */
+    BIO_ADDRINFO_free(ai);
+
+    /* Create a BIO to wrap the socket*/
+    bio = BIO_new(BIO_s_socket());
+    if (bio == NULL)
+        BIO_closesocket(sock);
+
+    /*
+     * Associate the newly created BIO with the underlying socket. By
+     * passing BIO_CLOSE here the socket will be automatically closed when
+     * the BIO is freed. Alternatively you can use BIO_NOCLOSE, in which
+     * case you must close the socket explicitly when it is no longer
+     * needed.
+     */
+    BIO_set_fd(bio, sock, BIO_CLOSE);
+
+    return bio;
+}
+
+/* Server hostname and port details */
+#define HOSTNAME "www.example.com"
+#define PORT     "443"
+
+/*
+ * Simple application to send a basic HTTP/1.0 request to a server and
+ * print the response on the screen.
+ */
+int main(void)
+{
+    SSL_CTX *ctx = NULL;
+    SSL *ssl;
+    BIO *bio = NULL;
+    int res = EXIT_FAILURE;
+    int ret;
+    const char *request = "GET / HTTP/1.0\r\nHost: "HOSTNAME"\r\n\r\n";
+    size_t written, readbytes;
+    char buf[160];
+
+    /*
+     * Create an SSL_CTX which we can use to create SSL objects from. We
+     * want an SSL_CTX for creating clients so we use TLS_client_method()
+     * here.
+     */
+    ctx = SSL_CTX_new(TLS_client_method());
+    if (ctx == NULL) {
+        printf("Failed to create the SSL_CTX\n");
```

Comment:
I would use a wrapper function that prints the error stack.

Reply:
I'm not sure it adds much? The error stack printing is already there and in the common "end" block. The only thing different about each location is printing the error message itself.